speed.py

```diff
@@ -0,0 +1,99 @@
+from datasets import load_dataset
+from collections import defaultdict
+import matplotlib.pyplot as plt
+import pandas as pd
+import numpy as np
+
+with load_dataset("roszcz/internship-midi-data-science", split="train") as dataset:
+    record = dataset[0]
+    df = (pd.DataFrame(record["notes"])).sort_values(by='start')
+
+def get_events_frequency(events_df):
+    """
+    Calculate the frequency of events in 60-second intervals.
+
+    Parameters:
+        events_df (DataFrame): A pandas DataFrame containing events with 'start' and 'end' columns.
+
+    Returns:
+        DataFrame: A new DataFrame with the frequency of events in each 60-second interval.
+
+    Example:
+        events_df = pd.DataFrame({
+            'start': [0, 30, 45, 80],
+            'end': [15, 60, 50, 120]
+        })
+        result_df = get_events_frequency(events_df)
+        print(result_df)
+
+        Output:
+           minute     speed
+        0       0  0.250000
+        1       1  0.750000
+        2       2  0.750000
+        3       3  0.750000
+        4       4  0.750000
+        5       5  0.750000
+        6       6  0.750000
+        7       7  0.750000
+        8       8  0.750000
+        9       9  0.750000
+        10     10  0.750000
+        11     11  0.750000
+        12     12  0.750000
+        13     13  0.750000
+        14     14  0.750000
+        15     15  0.750000
+        16     16  0.750000
+        17     17  0.750000
+        18     18  0.750000
+        19     19  0.750000
+        20     20  0.750000
+        21     21  0.750000
+        22     22  0.750000
+        23     23  0.750000
+    """
+
+    # Sort events based on their start times
+    sorted_events = events_df.sort_values(by='start')
+
+    # Create a dictionary to store the count of events in each 60-second interval
+    interval_counts = defaultdict(int)
+
+    # Iterate through the sorted events and count the events in each interval
+    for index, event in sorted_events.iterrows():
+        start_time = event['start']
+        end_time = event['end']
+
+        # Determine the start and end intervals for the event
+        start_interval = start_time // 60
+        end_interval = (end_time - 1) // 60

```

Comment:
I'm not sure if the "interval" is the best name here - why did you choose it? As far as I can tell it's the rounded down minute of record. Can you also explain the `-1` in `end_interval`?

Reply:
The interval: 
My other options were chunk,period,and a minute-i wanted to avoid using period,chunk is not informative,and a minute i considered to restrict the scale,as there might appear a need to use another time unit
The -1: 
Using end_interval - 1 ensures that the note is only counted in the correct interval and not the next one, so that it is not doubled

Reply:
I'm still not sure if I follow the logic here, here's more questions:

1. Looks like `end_interval` can only be the same as `start_interval` or smaller. What's going to happen for a note with `start_time=120.1, end_time=120.2`?
2. Other than the case from 1., I think all of the notes for every interval will fall into the same bucket. Is there any difference between what you are doing, and something like this:

```python
# Could be a parameter
time_interval = 60
df["time_bin"] = df.start // time_interval
```

Reply:
1: I have tested it with or without substraction,also scaling time up so that there would be no such cases possible,
and results remain the same for visualization
Main reason are the chords/sequences segmentation 
It will supposedly ignore occurences of rare/unique fast chords,but preserve long chords, chords progressions,and common triads or dyads
Also it will supposedly preserve extended triads even if they are relatively rare
2: You are right,it makes absolutely no difference,essentially the bins' method is shown in the option with histogram
I thought that it would make the operations on start and end more readable,and easier to change the code for other tasks